### PR TITLE
owner, changefeed: backport async sink to old owner

### LINKED ISSUE #2295: Solve sink failure interrupting replication in Old Owner
Irresponsive downstreams will block the owner until the operation times out. For example, if the Kafka broker is down and the owner tries to send a checkpoint event, the owner would not be able to do ANYTHING until the Kafka sink has timed out (if ever). This will only happen in the old owner. 

# Action item

- [x] Back port `asyncSink` to the old owner
- [x] Make old owner's DDL execution asynchronous (in `changefeed.go`)

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve? 
close #2295  
### What is changed and how it works?

- [x] Back port `asyncSink` to the old owner
- [x] Make old owner's DDL execution asynchronous (in `changefeed.go`)
### Check List 

Tests 

 - Unit test
 - integration test 
 - Manual test (add detailed scripts or steps below)

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity

### Release note 

```release-note
None
```

### CODE REVIEW

File: cdc/async_sink.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	cerror "github.com/pingcap/ticdc/pkg/errors"
+	"github.com/pingcap/ticdc/pkg/filter"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultErrChSize = 1024
+)
+
+// AsyncSink is a async sink design for owner
```
Comment (amyangfei) [inline cdc/async_sink.go:37]:
```suggestion
// AsyncSink is an async sink design for owner
```

File: cdc/async_sink.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdc
```
Comment (overvenus) [inline cdc/async_sink.go:14]: Please add a TODO to remove this file once new owner is enabled.

File: cdc/owner.go [hunk 3]
```diff
@@ -428,15 +429,31 @@ func (o *Owner) newChangeFeed(
 	}
 	errCh := make(chan error, 1)
 
-	primarySink, err := sink.NewSink(ctx, id, info.SinkURI, filter, info.Config, info.Opts, errCh)
+	// primarySink, err := sink.NewSink(ctx, id, info.SinkURI, filter, info.Config, info.Opts, errCh)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
```
Comment (overvenus) [inline cdc/owner.go:435]:
```suggestion
```

File: cdc/async_sink.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	// TODO make the tick duration configurable
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	var lastCheckpointTs model.Ts
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-s.errCh:
+			ctx.Throw(err)
```
Comment (amyangfei) [inline cdc/async_sink.go:111]: Should we create a `throwContext`, otherwise the thrown error is not watched.

File: cdc/owner.go [hunk 5]
```diff
@@ -428,15 +429,31 @@ func (o *Owner) newChangeFeed(
...
+	// new asyncSink
+	cdcCtx := cdcContext.NewContext(ctx, &cdcContext.GlobalVars{})
+	cdcCtx = cdcContext.WithChangefeedVars(cdcCtx, &cdcContext.ChangefeedVars{
+		ID:   id,
+		Info: info,
+	})
+	primaryAsyncSink, err := newAsyncSink(cdcCtx)
```
Comment (overvenus) [inline cdc/owner.go:442]: What about using cdc/owner/async_sink directly?

File: cdc/async_sink.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+			if err := s.sink.EmitCheckpointTs(ctx, checkpointTs); err != nil {
+				ctx.Throw(errors.Trace(err))
+				return
+			}
+		case ddl := <-s.ddlCh:
+			err := s.sink.EmitDDLEvent(ctx, ddl)
+			failpoint.Inject("InjectChangefeedDDLError", func() {
+				err = cerror.ErrExecDDLFailed.GenWithStackByArgs()
+			})
+			if err == nil || cerror.ErrDDLEventIgnored.Equal(errors.Cause(err)) {
```
Comment (amyangfei) [inline cdc/async_sink.go:128]:
```suggestion
			if err == nil || cerror.ErrDDLEventIgnored.Equal(err) {
```

`Equal` will call `errors.Cause` on err

File: cdc/changefeed.go [hunk 7]
```diff
@@ -706,42 +726,34 @@ func (c *changeFeed) handleDDL(ctx context.Context) error {
 		c.ddlJobHistory = c.ddlJobHistory[1:]
 		c.ddlExecutedTs = todoDDLJob.BinlogInfo.FinishedTS
 		c.ddlState = model.ChangeFeedSyncDML
+		c.ddlEventCache = nil
 		return nil
 	}
 
-	executed := false
 	if !c.cyclicEnabled || c.info.Config.Cyclic.SyncDDL {
```
Comment (amyangfei) [inline cdc/changefeed.go:733]: c.ddlJobHistory, c.ddlExecutedTs and c.ddlState are not updated if `c.cyclicEnabled` is true and `c.info.Config.Cyclic.SyncDDL` is false

Comment (asddongmen) [reply cdc/changefeed.go:733]: thanks!

File: cdc/owner.go (hunk 5 above)
Comment (asddongmen) [reply cdc/owner.go:442]: I will try this.

File: cdc/async_sink.go (hunk 2 above)
Comment (asddongmen) [reply cdc/async_sink.go:14]: Ok, but seems that all old owner related code should be cleaned while old owner is enabled.

File: cdc/changefeed.go [hunk 8]
```diff
@@ -706,42 +726,34 @@ func (c *changeFeed) handleDDL(ctx context.Context) error {
...
+			c.ddlJobHistory = c.ddlJobHistory[1:]
+			c.ddlExecutedTs = todoDDLJob.BinlogInfo.FinishedTS
+			c.ddlState = model.ChangeFeedSyncDML
+			c.ddlEventCache = nil
 		}
 	}
-	if executed {
-		log.Info("Execute DDL succeeded", zap.String("changefeed", c.id), zap.Reflect("ddlJob", todoDDLJob))
-	} else {
-		log.Info("Execute DDL ignored", zap.String("changefeed", c.id), zap.Reflect("ddlJob", todoDDLJob))
```
Comment (liuzix) [inline cdc/changefeed.go:739]: We need to preserve this log

File: cdc/changefeed.go [hunk 9]
```diff
@@ -706,42 +726,34 @@ func (c *changeFeed) handleDDL(ctx context.Context) error {
...
 		failpoint.Inject("InjectChangefeedDDLError", func() {
 			failpoint.Return(cerror.ErrExecDDLFailed.GenWithStackByArgs())
 		})
 
 		ddlEvent.Query = binloginfo.AddSpecialComment(ddlEvent.Query)
 		log.Debug("DDL processed to make special features mysql-compatible", zap.String("query", ddlEvent.Query))
-		err = c.sink.EmitDDLEvent(ctx, ddlEvent)
+
+		c.ddlEventCache = ddlEvent
+		cdcCtx := cdcContext.NewContext(ctx, &cdcContext.GlobalVars{})
```
Comment (liuzix) [inline cdc/changefeed.go:742]: It is better if we do not create the `cdcCtx` afresh each time.

File: cdc/changefeed.go (hunk 8 above)
Comment (asddongmen) [reply cdc/changefeed.go:739]: ok

File: cdc/owner.go (hunk 5 above)
Comment (asddongmen) [reply cdc/owner.go:442]: I made some changes on async_sink, so I think it’s better to use it separately from `cdc/owner/async_sink`.

File: cdc/changefeed.go (hunk 8 above)
Comment (asddongmen) [reply cdc/changefeed.go:739]: Thes log will print  in `async_sink`.

File: cdc/changefeed.go (hunk 7 above)
Comment (asddongmen) [reply cdc/changefeed.go:734]: I have fix these code.
Please help me to review it again.

File: cdc/changefeed.go (hunk 9 above)
Comment (asddongmen) [reply cdc/changefeed.go]: done

File: cdc/async_sink.go (hunk 6 above)
Comment (asddongmen) [reply cdc/async_sink.go]: done

File: cdc/async_sink.go (hunk 1 above)
Comment (asddongmen) [reply cdc/async_sink.go]: done

File: cdc/owner.go (hunk 3 above)
Comment (asddongmen) [reply cdc/owner.go]: thanks! done.

File: cdc/async_sink.go (hunk 2 above)
Comment (asddongmen) [reply cdc/async_sink.go:14]: done

File: cdc/changefeed.go (hunk 8 above)
Comment (liuzix) [reply cdc/changefeed.go:739]: Please make sure that these logs will be printed exactly the same way, or otherwise `ddl_reentrant` test will not pass.

Comment (asddongmen) [reply cdc/changefeed.go:739]: thank you very much~!

File: cdc/async_sink.go (hunk 4 above)
Comment (asddongmen) [reply cdc/async_sink.go:114]: done

File: cdc/owner.go [hunk 10]
```diff
@@ -426,17 +428,31 @@ func (o *Owner) newChangeFeed(
...
-	primarySink, err := sink.NewSink(ctx, id, info.SinkURI, filter, info.Config, info.Opts, errCh)
+	cdcCtx = cdcContext.WithChangefeedVars(cdcCtx, &cdcContext.ChangefeedVars{
+		ID:   id,
+		Info: info,
+	})
+	errCh := make(chan error, 1)
+	cdcCtx = cdcContext.WithErrorHandler(cdcCtx, func(err error) error {
+		errCh <- errors.Trace(err)
+		return nil
+	})
```
Comment (amyangfei) [inline cdc/owner.go:440]: This `errCh` is consumed only once, could it happen multiple errors need to be thrown and error sending blocks here

**Comment by overvenus** [comment]
PTAL @leoppro

File: cdc/owner.go (hunk 10 above)
Comment (asddongmen) [reply cdc/owner.go:440]: thanks

Comment (asddongmen) [reply cdc/owner.go:440]: done

**Comment by liuzix** [comment]
Why is the integration test case in `ddl_reentrant`?

**Comment by asddongmen** [comment]
rnu-integration-tests

**Review by leoppro — COMMENTED** [review]
LGTM

**Review by overvenus — APPROVED** [review]

**Pull request merged**